x/warp/contract.go

```diff
@@ -0,0 +1,277 @@
+type WarpMessage struct {
+	OriginChainID       [32]byte
+	OriginSenderAddress [32]byte
+	DestinationChainID  [32]byte
+	DestinationAddress  [32]byte
+	Payload             []byte
+}
+
+type GetVerifiedWarpMessageOutput struct {
+	Message WarpMessage
+	Exists  bool
+}
+
+type SendWarpMessageInput struct {
+	DestinationChainID [32]byte
+	DestinationAddress [32]byte
+	Payload            []byte
+}
+
+// PackGetBlockchainID packs the include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
+func PackGetBlockchainID() ([]byte, error) {
+	return WarpABI.Pack("getBlockchainID")
+}
+
+// PackGetBlockchainIDOutput attempts to pack given blockchainID of type [32]byte
+// to conform the ABI outputs.
+func PackGetBlockchainIDOutput(blockchainID [32]byte) ([]byte, error) {
+	return WarpABI.PackOutput("getBlockchainID", blockchainID)
+}
+
+// getBlockchainID returns the snow Chain Context ChainID of this blockchain.
+func getBlockchainID(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = contract.DeductGas(suppliedGas, GetBlockchainIDGasCost); err != nil {
+		return nil, 0, err
+	}
+	packedOutput, err := PackGetBlockchainIDOutput(accessibleState.GetSnowContext().ChainID)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// PackGetVerifiedWarpMessage packs the include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
+func PackGetVerifiedWarpMessage() ([]byte, error) {
+	return WarpABI.Pack("getVerifiedWarpMessage")
+}
+
+// PackGetVerifiedWarpMessageOutput attempts to pack given [outputStruct] of type GetVerifiedWarpMessageOutput
+// to conform the ABI outputs.
+func PackGetVerifiedWarpMessageOutput(outputStruct GetVerifiedWarpMessageOutput) ([]byte, error) {
+	return WarpABI.PackOutput("getVerifiedWarpMessage",
+		outputStruct.Message,
+		outputStruct.Exists,
+	)
+}
+
+// getVerifiedWarpMessage retrieves the pre-verified warp message from the predicate storage slots and returns
+// the expected ABI encoding of the message to the caller.
+func getVerifiedWarpMessage(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, _ []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	remainingGas = suppliedGas
+	// XXX Note: there is no base cost for retrieving a verified warp message. Instead, we charge for each piece of gas,
+	// prior to each execution step.
+	// Ignore input since there are no arguments
+	predicateBytes, exists := accessibleState.GetStateDB().GetPredicateStorageSlots(ContractAddress)
+	// If there is no such value, return false to the caller.
+	// Note: decoding errors will return an error instead.
```

Comment:
```suggestion
	// Note: encoding errors will return an error instead.
```

Reply:
This is correct as is. If there is an error decoding the message (as opposed to the message is not present), then the predicate will fail verification.

Reply:
nit: could consider moving this Note closer to the error being returned.

Reply:
Moved to right above `UnpackPredicate`